---
title: "GSEA"
author: "Princess Rodriguez"
date: "4/15/2024"
output:
  html_document:
    toc: yes
    toc_depth: 4
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, warning=FALSE, message=FALSE}
library(clusterProfiler)
library(dplyr)
library(msigdbr) # BiocManager::install("msigdb")
library(pathview)
library(enrichplot) # BiocManager::install("enrichplot")
```

## Other methods for functional analysis

Over-representation analysis is only a single type of functional analysis method that is available for teasing apart the biological processes important to your condition of interest. Other types of analyses can be equally important or informative, including functional class scoring and pathway topology methods. 

![Pathway analysis tools](img/pathway_analysis.png)

## Functional class scoring tools

Functional class scoring (FCS) tools, such as [GSEA](http://software.broadinstitute.org/gsea/index.jsp), most often use the gene-level statistics or log2 fold changes for all genes from the differential expression results, then look to see whether gene sets for particular biological pathways are enriched among the large positive or negative fold changes. 

![GSEA example](img/gsea_theory.png)

The hypothesis of FCS methods is that although large changes in individual genes can have significant effects on pathways (and will be detected via ORA methods), weaker but coordinated changes in sets of functionally related genes (i.e., pathways) can also have significant effects.  Thus, rather than setting an arbitrary threshold to identify 'significant genes', **all genes are considered** in the analysis. The gene-level statistics from the dataset are aggregated to generate a single pathway-level statistic and statistical significance of each pathway is reported. This type of analysis can be particularly helpful if the differential expression analysis only outputs a small list of significant DE genes. 

### Preparation for GSEA

Using the log2 fold changes obtained from the DE analysis for every gene, gene set enrichment analysis (GSEA) can be performed using clusterProfiler.

To perform GSEA analysis of KEGG gene sets, clusterProfiler can be used, but requires the genes to be identified using Entrez IDs. We also need to remove the NA values and duplicates (due to gene ID conversion) prior to the analysis. 

```{r}
gsea_data <- read.csv("TCF1KOvsWT_all_matrix_entrez.csv", header = TRUE)

#gsea_data
```



```{r}
## Remove any NA values 
res_entrez <- dplyr::filter(gsea_data, ENTREZID != "NA")
```

+ The `::` operator is used to access functions or objects from specific packages. The `::` operator is used to specify that the `filter()` function is being called from the `dplyr` package. This is a way to explicitly specify the package namespace for the function to avoid conflicts with functions of the same name from other packages that might be loaded into the R environment.

+ In programming, `!=` is a comparison operator that stands for "not equal to." So above, the expression `ENTREZID != "NA"` means "select rows from the gsea_data data frame where the value in the ENTREZID column is not equal to the string "NA"."


```{r}
## Remove any Entrez duplicates
res_entrez <- res_entrez[which(duplicated(res_entrez$ENTREZID) == F), ]
```

1.  `duplicated(res_entrez$ENTREZID)` returns a logical vector indicating whether each element in the `ENTREZID` column is duplicated or not.

2. `which(duplicated(res_entrez$ENTREZID) == FALSE)` returns the indices of elements in the logical vector where the value is FALSE, i.e., where the ENTREZID values are not duplicated.

  + logical_vector <- c(TRUE, FALSE, TRUE, TRUE, FALSE)
  + indices <- which(logical_vector == TRUE)
    + [1] 1 3 4

3. `res_entrez[which(duplicated(res_entrez$ENTREZID) == FALSE), ]` subsets the res_entrez data frame, keeping only the rows where the ENTREZID values are not duplicated.

Finally, we will extract and name the fold changes:

```{r}
## Extract the foldchanges
foldchanges <- res_entrez$log2FoldChange

## Name each fold change with the corresponding Entrez ID
names(foldchanges) <- res_entrez$ENTREZID
```


```{r}
head(foldchanges)
```


Next we need to order the fold changes in decreasing order. To do this we'll use the `sort()` function, which takes a vector as input. This is in contrast to Tidyverse's `arrange()`, which requires a data frame.

```{r}
## Sort fold changes in decreasing order
foldchanges <- sort(foldchanges, decreasing = TRUE)

head(foldchanges)
```


```{r}
#saveRDS(foldchanges, "foldchanges.RDS")

#foldchanges <- readRDS("foldchanges.RDS")
```


### Theory of GSEA

Now we are ready to perform GSEA. The details regarding GSEA can be found in the [PNAS paper](https://www.pnas.org/content/102/43/15545) by Subramanian et al. We will describe briefly the steps outlined in the paper below:

![GSEA overview](img/gsea_overview.png)


_**Image credit:** [Subramanian et al. Proceedings of the National Academy of Sciences Oct 2005, 102 (43) 15545-15550; DOI: 10.1073/pnas.0506580102](https://www.pnas.org/content/102/43/15545)_

This image describes the theory of GSEA, with the 'gene set S' showing the metric used (in our case, ranked log2 fold changes) to determine enrichment of genes in the gene set. The left-most image is representing this metric used for the GSEA analysis. The log2 fold changes for each gene in the 'gene set S' is shown as a line in the middle image. The large positive log2 fold changes are at the top of the gene set image, while the largest negative log2 fold changes are at the bottom of the gene set image. In the right-most image, the gene set is turned horizontally, underneath which is an image depicting the calculations involved in determining enrichment, as described below.

**Step 1:** Calculation of enrichment score: 

An **enrichment score** for a particular gene set is calculated by walking down the list of log2 fold changes and increasing the running-sum statistic every time a gene in the gene set is encountered and decreasing it when genes are not part of the gene set. The size of the increase/decrease is determined by magnitude of the log2 fold change. Larger (positive or negative) log2 fold changes will result in larger increases or decreases. The final enrichment score is where the running-sum statistic is the largest deviation from zero.

**Step 2:** Estimation of significance:

The **significance** of the enrichment score is determined using permutation testing, which performs rearrangements of the data points to determine the likelihood of generating an enrichment score as large as the enrichment score calculated from the observed data. Essentially, for this step, the first permutation would reorder the log2 fold changes and randomly assign them to different genes, reorder the gene ranks based on these new log2 fold changes, and recalculate the enrichment score. The second permutation would reorder the log2 fold changes again and recalculate the enrichment score again, and this would continue for the total number of permutations run. Therefore, the number of permutations run will increase the confidence in the signficance estimates. The default number of times this permutation test is run is 1000X! 

**Step 3:** Adjust for multiple test correction

After all gene sets are tested, the enrichment scores are normalized for the size of the gene set, then the p-values are corrected for multiple testing.

The GSEA output will yield the core genes in the gene sets that most highly contribute to the enrichment score. The genes output are generally the genes at or before the running sum reaches its maximum value (eg. the most influential genes driving the differences between conditions for that gene set).

GSEA statistics:

+ Calculation of an Enrichment Score.
The primary result of the gene set enrichment analysis is the enrichment score (ES), which reflects the degree to which a gene set is overrepresented at the top or bottom of a ranked list of genes. GSEA calculates the ES by walking down the ranked list of genes, increasing a running-sum statistic when a gene is in the gene set and decreasing it when it is not. The magnitude of the increment depends on the correlation of the gene with the phenotype. The ES is the maximum deviation from zero encountered in walking the list. A positive ES indicates gene set enrichment at the top of the ranked list; a negative ES indicates gene set enrichment at the bottom of the ranked list.

+ Normalized Enrichment Score. 
The normalized enrichment score (NES) is the primary statistic for examining gene set enrichment results. By normalizing the enrichment score, GSEA accounts for differences in gene set size and in correlations between gene sets and the expression dataset; therefore, the normalized enrichment scores (NES) can be used to compare analysis results across gene sets.

+ False Discovery Rate (FDR)
The false discovery rate (FDR) is the estimated probability that a gene set with a given NES represents a false positive finding. An FDR cutoff of 5% is standard. 

+ Nominal P Value
The nominal p value estimates the statistical significance of the enrichment score for a single gene set.

### Performing GSEA

Within clusterprofiler we can take advantage of the `gseGO` and `gseKEGG` function which will allow us to perform Gene Set Enrichment Analysis.  

**DO NOT RUN!!!** 

Class is like a cooking show sometimes, I will have objects prepared in advanced so we can see and interpret results together. At home, you will find it takes longer!

```{r}
## GSEA using gene sets from KEGG pathways
gseaKEGG <- gseKEGG(geneList = foldchanges, # ordered named vector of fold changes 
                    organism = "mmu", # supported organisms listed below
                    nPerm = 1000, # default number permutations
                    minGSSize = 20, # minimum gene set size (# genes in set)
                    pvalueCutoff = 0.05, # padj cutoff value
                    verbose = FALSE)
```


```{r}
#saveRDS(gseaKEGG, file = "gseaKEGG.RDS")

#gseaKEGG <- readRDS("gseaKEGG.RDS")
```


```{r}
## Extract the GSEA results
gseaKEGG_results <- gseaKEGG@result
```


The organisms with KEGG pathway information are listed [here](http://www.genome.jp/kegg/catalog/org_list.html).

**How many pathways are enriched?** 

```{r}
## Write GSEA results to file
write.csv(gseaKEGG_results, "gsea_kegg_results.csv", quote=F)
```

**NOTE:** We will all get different results for the GSEA because the permutations performed use random reordering. If we would like to use the same permutations every time we run a function (i.e. we would like the same results every time we run the function), then we could use the `set.seed(123456)` function prior to running. The input to `set.seed()` could be any number, but if you would want the same results, then you would need to use the same number as input.

View the enriched pathways:
```{r}
#View(gseaKEGG_results)
```

* The first few columns of the results table identify the pathway information
* The following columns include the associated statistics
* The last column will report which genes are part of the 'core enrichment'. These are the genes associated with the pathway which contributed to the observed enrichment score (i.e. in the extremes of the ranking). The genes are listed by EntrezID.


Now lets explore the GSEA plot of enrichment for one of the pathways in the ranked list:

```{r, fig.width=6, fig.height=4}
gsea1 <- gseaplot(gseaKEGG, title = gseaKEGG_results$Description[1], geneSetID = 1)
gsea1
```

Now lets view this using `gseaplot2()` from the `library(enrichplot)`

```{r, fig.width=6, fig.height=4}
enrich_gsea <- gseaplot2(gseaKEGG, 
                         title = gseaKEGG_results$Description[1], 
                         geneSetID = 1)
enrich_gsea
```

Original Hypothesis: The depletion of TCF7 in CD8+ T cells will lead to an enrichment of genes involved in myeloid cell development.

```{r, fig.width=6, fig.height=4}
gseaKEGG_top5 <- gseaplot2(gseaKEGG, geneSetID = 1:5)
gseaKEGG_top5
```



Use the [Pathview R package](http://bioconductor.org/packages/release/bioc/html/pathview.html) to integrate the KEGG pathway data from clusterProfiler into pathway images:

```{r}
#detach("package:dplyr", unload=TRUE) # first unload dplyr to avoid conflicts

## Output images for a single significant KEGG pathway
pathview(gene.data = foldchanges,
              pathway.id = "mmu04640",
              species = "mmu",
              limit = list(gene = 2, # value gives the max/min limit for foldchanges
              cpd = 1))

knitr::include_graphics("mmu04640.pathview.png")
```

>**NOTE:** If the below error message occurs: `Error in detach("package:dplyr", unload = T) : invalid 'name' argument`, that means the dplyr package is not currently loaded. Ignore the message and continue to run pathview command.


### Incorpororating other gene sets for GSEA

The Molecular Signatures Database (also known as [MSigDB](http://software.broadinstitute.org/gsea/msigdb/index.jsp)) is a collection of annotated gene sets. It contains 9 major collections:

* H: hallmark gene sets
* C1: positional gene sets
* C2: curated gene sets
* C3: motif gene sets
* C4: computational gene sets
* C5: GO gene sets
* C6: oncogenic signatures
* C7: immunologic signatures
* C8: cell type signature gene sets 

Users can download **GMT files** from Broad Institute and use the `read.gmt()` function to parse the files. Alternatively, there is an R package that already packed the MSigDB gene sets in tidy data format that can be used directly with clusterProfiler. 

```{r}
#library(msigdbr)
#install.packages("msigdbr")
```

Available functions 

`msigdbr()`
    Retrieve the gene sets data frame

`msigdbr_collections()`
    List the collections available in the msigdbr package

`msigdbr_species()`
    List the species available in the msigdbr package 


MSigDB contains gene sets **only for human**, but it can be extended to other organisms by mapping to the homologues genes. The package msigdbr has already mapped genes to many other organisms. A full list of supported organisms in msigdbr can be obtained by：

```{r}
msigdbr_species()
```


To obtain all gene sets in the database for a species, use the `msigdbr()` function and specify a species of interest.
```{r}
#all_gene_sets <- msigdbr(species = "Mus musculus")

#saveRDS(all_gene_sets, file = "all_gene_sets.RDS")
```

```{r}
all_gene_sets <- readRDS(file = "all_gene_sets.RDS")
```

```{r}
head(all_gene_sets)
```


```{r}
msigdbr_collections()
```

To use a specific collection; example is C5:GOBP
```{r}
GOBP_gene_sets <- msigdbr(species = "mouse", category = "C5",
                            subcategory = "GO:BP") %>% 
  dplyr::select(gs_name, entrez_gene)

head(GOBP_gene_sets)
```

```{r}
#saveRDS(GOBP_gene_sets, "GOBP_gene_sets.RDS")

#GOBP_gene_sets <- readRDS("GOBP_gene_sets.RDS")
```


```{r}
# Run GSEA
#msig_GSEA <- GSEA(ordered rank gene list, 
                  #TERM2GENE = dataframe of 2 columns with term and gene,
                  #verbose = FALSE)
```


**This can take a few minutes to run, DO NOT RUN**

```{r}
# Run GSEA
#msig_GOBP_GSEA <- GSEA(foldchanges, TERM2GENE = GOBP_gene_sets, verbose = FALSE)
```


```{r}
#saveRDS(msig_GOBP_GSEA, "msig_GOBP_GSEA.RDS")

msig_GOBP_GSEA <- readRDS("msig_GOBP_GSEA.RDS")
```


```{r}
## Extract the GSEA results
gseaGOBP_results <- msig_GOBP_GSEA@result
```

```{r}
## Write GSEA results to file
write.csv(gseaGOBP_results, "gseaGOBP_results.csv", quote=F)
```


```{r}
#View(gseaGOBP_results)
```


Original Hypothesis: The depletion of TCF7 in CD8+ T cells will lead to an enrichment of genes involved in myeloid cell development.


```{r, fig.height=4, fig.width=6}
## Plot the GSEA plot for a single enriched pathway
gseaplot2(msig_GOBP_GSEA, title = 'GOBP_MYELOID_LEUKOCYTE_ACTIVATION', geneSetID = 'GOBP_MYELOID_LEUKOCYTE_ACTIVATION')
```




```{r}
#BiocManager::install("GSEABase")
#library(GSEABase)

# Load in GMT file of gene sets (we downloaded from the Broad Institute for MSigDB)

#c2 <- read.gmt("/data/c2.cp.v6.0.entrez.gmt.txt")

#msig <- GSEA(foldchanges, TERM2GENE=c2, verbose=FALSE)

#msig_df <- data.frame(msig)
```




## Resources for functional analysis

* g:Profiler - http://biit.cs.ut.ee/gprofiler/index.cgi 
* DAVID - http://david.abcc.ncifcrf.gov/tools.jsp 
* clusterProfiler - http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html
* GeneMANIA - http://www.genemania.org/
* GenePattern -  http://www.broadinstitute.org/cancer/software/genepattern/ (need to register)
* WebGestalt - http://bioinfo.vanderbilt.edu/webgestalt/ (need to register)
* AmiGO - http://amigo.geneontology.org/amigo
* ReviGO (visualizing GO analysis, input is GO terms) - http://revigo.irb.hr/ 
* WGCNA - https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/
* GSEA - http://software.broadinstitute.org/gsea/index.jsp
* SPIA - https://www.bioconductor.org/packages/release/bioc/html/SPIA.html
* GAGE/Pathview - http://www.bioconductor.org/packages/release/bioc/html/gage.html

***
*This lesson has been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*
